src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisResultBuilder.cs

```diff
@@ -95,14 +101,51 @@ public TimeSpan GetAnalyzerExecutionTime(DiagnosticAnalyzer analyzer)
             }
         }
 
-        internal ImmutableArray<DiagnosticAnalyzer> GetPendingAnalyzers(ImmutableArray<DiagnosticAnalyzer> analyzers)
+        private HashSet<DiagnosticAnalyzer>? GetCompletedAnalyzersForFile_NoLock(SourceOrAdditionalFile filterFile, bool syntax)
+        {
+            if (filterFile.SourceTree is { } tree)
+            {
+                var completedAnalyzersByTree = syntax ? _completedSyntaxAnalyzersByTree : _completedSemanticAnalyzersByTree;
+                if (completedAnalyzersByTree.TryGetValue(tree, out var completedAnalyzersForTree))
+                {
+                    return completedAnalyzersForTree;
+                }
+            }
+            else if (_completedSyntaxAnalyzersByAdditionalFile.TryGetValue(filterFile.AdditionalFile!, out var completedAnalyzersForFile))
+            {
+                return completedAnalyzersForFile;
+            }
+
+            return null;
+        }
+
+        private void AddCompletedAnalyzerForFile_NoLock(SourceOrAdditionalFile filterFile, bool syntax, DiagnosticAnalyzer analyzer)
+        {
+            var completedAnalyzers = new HashSet<DiagnosticAnalyzer> { analyzer };
+            if (filterFile.SourceTree is { } tree)
+            {
+                var completedAnalyzersByTree = syntax ? _completedSyntaxAnalyzersByTree : _completedSemanticAnalyzersByTree;
+                completedAnalyzersByTree.Add(tree, completedAnalyzers);
+            }
+            else
+            {
+                _completedSyntaxAnalyzersByAdditionalFile.Add(filterFile.AdditionalFile!, completedAnalyzers);
+            }
+        }
+
+        internal ImmutableArray<DiagnosticAnalyzer> GetPendingAnalyzers(ImmutableArray<DiagnosticAnalyzer> analyzers, (SourceOrAdditionalFile file, bool syntax)? filterScope = null)
         {
             lock (_gate)
             {
+                var completedAnalyzersForFile = filterScope.HasValue
+                    ? GetCompletedAnalyzersForFile_NoLock(filterScope.Value.file, filterScope.Value.syntax)
+                    : null;
+
                 ArrayBuilder<DiagnosticAnalyzer>? builder = null;
```

Comment:
fwiw, i've never found this pattern of avoiding the builder to help perf in any way.